Automattic/jetpack
# Jetpack CRM: Use WP_UnitTestCase instead of WorDBless

Jetpack CRM's codebase is a bit messy, and when trying to use WorDBless we sometimes end up fighting with different globally defined logic and assumptions that wasn't made with tests in mind.
We, therefore, want to start using [WP_UnitTestCase](https://make.wordpress.org/core/handbook/testing/automated-testing/phpunit/) instead like WordPress.org.

(Additionally: it's only a bonus that we will be able to do actual integration tests towards a "real" site to get away from a lot of the mocking we're doing to ensure our fragile legacy code doesn't break anything).

Fixes https://github.com/Automattic/zero-bs-crm/issues/3282

## Proposed changes:

* Update the test bootstrap logic [to follow the docs example](https://github.com/Automattic/jetpack/blob/trunk/docs/examples/bootstrap.php).
* Remove old `DBModeCheck` logic since we no longer have support for DAL 1 or 2 (our old data layer models).
    * The reason we remove it in this PR is an obscure race condition that happens when spinning up the test WordPress site.
* Add new parameter to `zeroBSCRM_database_reset` so we can bypass the permission check.
* Add a new Jetpack CLI command to run Jetpack CRM tests inside Docker (`jetpack docker phpunit-crm`)
* Update existing tests to use one of our new base case classes instead of `WorDBless\BaseTestCase`:
    * `JPCRM_Base_Test_Case`: ensures we do not let changed to the global `$zbs` logic bleed into other tests. 
    * `JPCRM_Base_Integration_Test_Case`: this will reset Jetpack CRM tables after each test so previous test data doesn't bleed into other tests.

### Other information:


## Jetpack product discussion



pbhBOz-3Uj-p2

## Does this pull request change what data or activity we track or use?



No

## Testing instructions:






If you're using the Docker environment from Jetpack Monorepo, then it should be as easy as running one command:

* Spin up the Jetpack Monorepo Docker environment
* Run: `jetpack docker phpunit-crm`
